### DISCUSSION TITLE
How to retrieve short/long option names (not values) of a parser?

### DISCUSSION THREAD

**Discussion by eugenesvk** [pacak/bpaf#395] [Q&A]
If I have 
```rs
let overwrite = short('o').long("overwrite").help("Overwrite target file").switch();
```

How do I retrieve `o` and `overwrite` options (**names**, not values) at runtime, for example, to be able to print some error message to the user `File exists, but overwrite flag isn't set, use '-o' or '--overwrite'` so that in case I later change `short('r')` to something else I wouldn't have to update the text of the error message?

Thanks!

**Comment by pacak** [comment]
At the moment there's no easy way, I'd have them as `const S_OVERWRITE: char = 'o';` and `const L_OVERWRITE ...` and use it both for the parser and for the error message.

Looking at hard ways - you can call `overwrite.meta()` and then recursively pattern match the result to get `ShortLong` that contains a visible names back. This will break once I publish a version I'm experimenting with now, but that version gives an easier way to access things.

**Reply by eugenesvk** [reply]
How would I get access to `overwrite.meta()` after I've used the `combine!` and `.to_options()`? `OptionsParser` doesn't seem to allow this, right?

`meta` seems to only be available to the `Parser` resulting from `combine!` before it's converted into runnable Options?

**Reply by pacak** [reply]
`.meta()` is accessible in any `Parser`, but not with `OptionParser`, right. `.meta()` will work though `construct!` just fine. If you have `OptionParser` - it will have all the options, not just `overwrite`. With `OptionParser` you could add some assertions (via completion mechanism, might require enabling) that flags `-o` and `--overwrite` are still present:


```rust
// assuming `options()` is your `OptionParser`
let r = options().run_inner(&Args::from(&["-o"]).set_comp(0)).unwrap_err().unwrap_stderr();
assert_eq!(r, "-o", "You need to update the error message");
let r = options().run_inner(&Args::from(&["--overwrite"]).set_comp(0)).unwrap_err().unwrap_stderr();
assert_eq!(r, "--overwrite", "You need to update the error message");
```

**Reply by eugenesvk** [reply]
But in the `OptionParser` case I wouldn't know what `"-o"` means, so it's not helpful to do the `run_inner`

Though actually that's the same issue with `Parser.meta()`: it has the flag names in the meta info, but doesn't include the struct field name (which is more permanent), so if in the future you add more conflicting options and reorder short/long options, you still wouldn't be able to auto-track which ones belong to the `overwrite` field

So I guess constants is indeed the only way now
